lib/cartopy/util.py

```diff
@@ -99,3 +101,298 @@ def add_cyclic_point(data, coord=None, axis=-1):
+    try:
+        slicer[axis] = slice(0, 1)
+    except IndexError:
+        estr = 'The specified axis does not correspond to an array dimension.'
+        raise ValueError(estr)
+    npc = np.ma if np.ma.is_masked(data) else np
+    return npc.concatenate((data, data[tuple(slicer)]), axis=axis)
+
+
+def _add_cyclic_lon(lon, axis=-1, cyclic=360):
+    """
+    Add a cyclic point to a longitude array.
+
+    Parameters
+    ----------
+    lon: ndarray, optional
+        An array which specifies the coordinate values for
+        the dimension the cyclic point is to be added to.
+    axis: int, optional
+        Specifies the axis of the longitude array to add the cyclic point to.
+        Defaults to the right-most axis.
+    cyclic: float, optional
+        Width of periodic domain (default: 360)
+
+    Returns
+    -------
+    The coordinate `lon` with a cyclic point added.
+    """
+    npc = np.ma if np.ma.is_masked(lon) else np
+    # get cyclic longitudes
+    # clon is the code from basemap (addcyclic)
+    # https://github.com/matplotlib/basemap/blob/master/lib/mpl_toolkits/basemap/__init__.py
+    clon = (np.take(lon, [0], axis=axis) +
+            cyclic * np.sign(np.diff(np.take(lon, [0, -1], axis=axis),
+                                     axis=axis)))
+    # basemap ensures that the values do not exceed cyclic
+    # (next code line). We do not do this to deal with rotated grids that
+    # might have values not exactly 0.
+    #     clon = npc.where(clon <= cyclic, clon, np.mod(clon, cyclic))
+    return npc.concatenate((lon, clon), axis=axis)
+
+
+def _has_cyclic(lon, axis=-1, cyclic=360, prec=1e-4):
+    """
+    Check if longitudes already have a cyclic point.
+
+    Checks all differences between the first and last
+    longitudes along `axis` to be less than `prec`.
+
+    Parameters
+    ----------
+    lon: ndarray, optional
+        An array with the coordinate values to be checked for cyclic points.
+    axis: int, optional
+        Specifies the axis of the `lon` array to be checked.
+        Defaults to the right-most axis.
+    cyclic: float, optional
+        Width of periodic domain (default: 360).
+    prec: float, optional
+        Maximal difference between first and last longitude to detect
+        cyclic point (default: 1e-4).
+
+    Returns
+    -------
+    True if a cyclic point was detected along the given axis,
+    False otherwise.
+    """
+    npc = np.ma if np.ma.is_masked(lon) else np
+    # transform to 0-cyclic, assuming e.g. -180 to 180 if any < 0
+    lon1 = np.mod(npc.where(lon < 0, lon + cyclic, lon), cyclic)
+    dd = np.diff(np.take(lon1, [0, -1], axis=axis), axis=axis)
+    if npc.all(np.abs(dd) < prec):
+        return True
+    else:
+        return False
+
+
+def add_cyclic(data, coord=None, rowcoord=None, axis=-1,
+               cyclic=360, prec=1e-4):
+    """
+    Add a cyclic point to an array and optionally corresponding
+    column (`coord` ~ longitudes) and row coordinates
```

Comment:
What about changing all `coord` to `xcoord`, and say that the wrapping can only occur in x? Then instead of `rowcoord` you could have `ycoord`.